ownCloud/Client/Actions/Actions+Extensions/UploadPhotosAction.swift

```diff
@@ -75,16 +79,52 @@ class UploadPhotosAction: UploadBaseAction {
 		}
 	}
 
-	func presentImageGalleryPicker() {
+	private func presentImageGalleryPicker() {
 		OnMainThread {
 			if let viewController = self.context.viewController {
 				let photoAlbumViewController = PhotoAlbumTableViewController()
 				photoAlbumViewController.selectionCallback = { (assets) in
-					for asset in assets {
-						self.upload(asset: asset)
-					}
 
 					self.completed()
+
+					let queue = DispatchQueue.global(qos: .userInitiated)
+
+					queue.async {
+
+						self.core?.perform(inRunningCore: { (runningCoreCompletion) in
+
+							let uploadGroup = DispatchGroup()
+							var uploadFailed = false
+
+							for asset in assets {
+								if uploadFailed == false {
+									// Upload image on a background queue
+									uploadGroup.enter()
+
+									self.upload(asset: asset, completion: { (success, uploadFinished) in
+										if !success {
+											uploadFailed = true
+										}
+										if uploadFinished {
+											uploadGroup.leave()
```

Comment:
(2/3) … but the group is only left when `self.upload(itemURL:…)` is called in L212.

The group is not left if there's f.ex. a conversion error…

Reply:
@felix-schwarz actually I made that change intentionally to leave the group only if actual upload has failed or succeeded to keep the OCCore alive as long as uploads are not finished. But actually.. yes, failed conversion was an open point. I fixed it with c78014a706c23a0254c370a78bce3c6de9f7a621

Reply:
Ok. If the core should be kept alive & open until the uploads have completed while *also* avoiding the blocking issues described in (3/3), another API may actually be better suited: call `OCCoreManager.requestCore()` to "retain" the core and `OCCoreManager.returnCore()` when you're done.